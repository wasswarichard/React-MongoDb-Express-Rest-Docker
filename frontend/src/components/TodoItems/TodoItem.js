import React from "react";
import {Typography, Button, Icon, Paper, Box, Checkbox,} from "@material-ui/core";
import * as config from '../../helpers/config'
import {store} from "../../state/store/store";
import {updateTodos} from "../../state/actions/TodosActions";
import {deleteTodos} from "../../state/actions/TodosActions";
import {makeStyles} from "@material-ui/core/styles";

const useStyles = makeStyles((theme) => ({
    todoContainer: {
        borderTop: "1px solid #bfbfbf",
        marginTop: 2,
        "&:first-child": {
            margin: 0,
            borderTop: "none",
        },
        "&:hover": {
            "& $deleteTodo": {
                visibility: "visible",
            },
        },
    },
    deleteTodo: {
        visibility: "hidden",
    },
}));


const TodoItem = ({todo, lastTodoElementRef, setValueChange}) => {
    const classes = useStyles();
    const authentication = store.getState().session
    function deleteTodo(todo) {
        fetch(`${config.apiUrl}/api/todo/${todo.todoId}`, {
            headers: {
                Accept: "application/json",
                "Content-Type": "application/json",
                "authorization": authentication.accessToken,
                "x-refresh": authentication.refreshToken
            },
            method: "DELETE",
        }).then((response) => {
            if(response.status === 403){
                localStorage.removeItem('session');
                window.location.href = '/';
            }
            setValueChange(Math.random() * 100)
            store.dispatch(deleteTodos(todo))

        })
    }
    function toggleTodoCompleted(todo) {
        fetch(`${config.apiUrl}/api/todo/${todo.todoId}`, {
            headers: {
                Accept: "application/json",
                "Content-Type": "application/json",
                "authorization": authentication.accessToken,
                "x-refresh": authentication.refreshToken
            },
            method: "PUT",
            body: JSON.stringify({
                ...todo,
                completed: !todo.completed
            }),
        }).then(response => {
                if (response.status === 403){
                    localStorage.removeItem('session');
                    window.location.href = '/';
                }
                setValueChange(Math.random() * 100)
                store.dispatch(updateTodos(todo));
        })
    }

    return (
        <Paper ref={lastTodoElementRef} key={todo._id} display="flex"  className={classes.todoContainer}>
            <Box display="flex" flexDirection="row" alignItems="center">
                <Checkbox checked={todo.completed} onChange={() => toggleTodoCompleted(todo)}/>
                <Box flexGrow={1}>
                    <Typography className={todo.completed ? classes.todoTextCompleted : ""} variant="body1">
                        {todo.text}
                    </Typography>
                </Box>
                <Box flexGrow={1}>
                    <Typography className={todo.completed ? classes.todoTextCompleted : ""} variant="body1">
                        { `Due Date:  ${new Date(todo.dueDate).toLocaleDateString()}`}
                    </Typography>
                </Box>
                <Button className={classes.deleteTodo} startIcon={<Icon>delete</Icon>} onClick={() => deleteTodo(todo)}>
                    Delete
                </Button>
            </Box>
        </Paper>

    )
}
export default TodoItem;