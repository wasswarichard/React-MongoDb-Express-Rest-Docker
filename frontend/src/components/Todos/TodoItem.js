import React, {useState} from "react";
import {Typography, Button, Icon, Paper, Box, Checkbox,} from "@material-ui/core";
import config from "../../helpers/config.json";
import {store} from "../../state/store/store";
import {updateTodos} from "../../state/actions/TodosActions";
import {deleteTodos} from "../../state/actions/TodosActions";

const TodoItem = ({todo, lastTodoElementRef, classes, setValueChange}) => {
    const authentication = store.getState().session

    function deleteTodo(todo) {
        fetch(`${config.apiUrl}/api/todo/${todo.todoId}`, {
            headers: {
                Accept: "application/json",
                "Content-Type": "application/json",
                "authorization": authentication.accessToken,
                "x-refresh": authentication.refreshToken
            },
            method: "DELETE",
        }).then((response) => {
            if(response.status === 200){
                setValueChange(Math.random() * 100)
                store.dispatch(deleteTodos(todo))
            }
        })
    }
    function toggleTodoCompleted(todo) {
        fetch(`${config.apiUrl}/api/todo/${todo.todoId}`, {
            headers: {
                Accept: "application/json",
                "Content-Type": "application/json",
                "authorization": authentication.accessToken,
                "x-refresh": authentication.refreshToken
            },
            method: "PUT",
            body: JSON.stringify({
                ...todo,
                completed: !todo.completed
            }),
        })
            .then(response => response.json())
            .then((todo) => {
                setValueChange(Math.random() * 100)
                store.dispatch(updateTodos(todo));
        });
    }

    return (
        <Paper ref={lastTodoElementRef} key={todo._id} display="flex" flexDirection="row" alignItems="center" className={classes.todoContainer}>
            <Box display="flex" flexDirection="row" alignItems="center">
                <Checkbox checked={todo.completed} onChange={() => toggleTodoCompleted(todo)}/>
                <Box flexGrow={1}>
                    <Typography className={todo.completed ? classes.todoTextCompleted : ""} variant="body1">
                        {todo.text}
                    </Typography>
                </Box>
                <Box flexGrow={1}>
                    <Typography className={todo.completed ? classes.todoTextCompleted : ""} variant="body1">
                        { `Due Date:  ${new Date(todo.dueDate).toLocaleDateString()}`}
                    </Typography>
                </Box>
                <Button className={classes.deleteTodo} startIcon={<Icon>delete</Icon>} onClick={() => deleteTodo(todo)}>
                    Delete
                </Button>
            </Box>
        </Paper>

    )
}
export default TodoItem;