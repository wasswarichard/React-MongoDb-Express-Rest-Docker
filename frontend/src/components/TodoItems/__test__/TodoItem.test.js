import React from "react";
import ReactDOM from "react-dom";
import TodoItem from "../TodoItem";
import { cleanup} from "@testing-library/react";
import TestRenderer from 'react-test-renderer';

afterEach(cleanup)
it('should render without crashing', function () {
    const div = document.createElement("div")
    const todo = {
        _id: 576897,
        completed: true
    }
    ReactDOM.render(<TodoItem todo={todo}/>, div)
});

it('should match snapshot', function () {
    const todo = {
        _id: 576897,
        completed: true
    }
    const tree = TestRenderer.create(<TodoItem todo={todo}/>).toJSON();
    expect(tree).toMatchSnapshot();

});
