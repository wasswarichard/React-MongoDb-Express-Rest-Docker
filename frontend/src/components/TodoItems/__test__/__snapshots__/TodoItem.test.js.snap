// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div
  className="MuiPaper-root makeStyles-todoContainer-1 MuiPaper-elevation1 MuiPaper-rounded"
  display="flex"
>
  <div
    className="MuiBox-root MuiBox-root-10"
  >
    <span
      aria-disabled={false}
      className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-4 MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-checked-5 Mui-checked MuiIconButton-colorSecondary"
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={null}
    >
      <span
        className="MuiIconButton-label"
      >
        <input
          checked={true}
          className="PrivateSwitchBase-input-7"
          data-indeterminate={false}
          onChange={[Function]}
          type="checkbox"
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
          />
        </svg>
      </span>
    </span>
    <div
      className="MuiBox-root MuiBox-root-11"
    >
      <p
        className="MuiTypography-root MuiTypography-body1"
      />
    </div>
    <div
      className="MuiBox-root MuiBox-root-12"
    >
      <p
        className="MuiTypography-root MuiTypography-body1"
      >
        Due Date:  Invalid Date
      </p>
    </div>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-deleteTodo-2"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        <span
          className="MuiButton-startIcon MuiButton-iconSizeMedium"
        >
          <span
            aria-hidden={true}
            className="material-icons MuiIcon-root"
          >
            delete
          </span>
        </span>
        Delete
      </span>
    </button>
  </div>
</div>
`;
